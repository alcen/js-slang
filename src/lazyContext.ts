import { Context } from './types'

export const LAZY_SOURCE_1 = 2001
export const LAZY_SOURCE_2 = 2002

/**
 * Checks whether the interpreter should evaluate
 * the expression lazily in this context. Currently,
 * lazy evaluation is only implemented for Source
 * chapters 1 and 2.
 *
 * @param context The context to be checked.
 * @returns True, if the interpreter is to evaluate
 *     lazily. False, if the interpreter should
 *     evaluate eagerly.
 */
export default function lazyEvaluate(context: Context): boolean {
  return lazyEvaluateInChapter(context.chapter)
}

/**
 * Checks if in the current context, the program will be
 * transpiled with the lazy evaluation transpiler.
 *
 * @param context The context to be checked.
 */
export function lazyEvaluateInTranspiler(context: Context): boolean {
  return lazyEvaluate(context) && context.executionMethod === 'native'
}

/**
 * Checks if in the current context, the program will be
 * interpreted with the lazy evaluation interpreter.
 *
 * @param context The context to be checked.
 */
export function lazyEvaluateInInterpreter(context: Context): boolean {
  return lazyEvaluate(context) && context.executionMethod === 'interpreter'
}

/**
 * Checks if we have to dynamically determine which execution
 * method to use in this context for lazy evaluation.
 *
 * @param context The context to be checked.
 */
export function lazyEvaluateAuto(context: Context): boolean {
  return lazyEvaluate(context) && context.executionMethod === 'auto'
}

/**
 * Checks whether the interpreter should evaluate the
 * expression lazily, given a chapter number.
 *
 * @param chapter The chapter to be checked.
 * @returns True, if the interpreter is to evaluate
 *     lazily. False, if the interpreter should
 *     evaluate eagerly.
 */
export function lazyEvaluateInChapter(chapter: number): boolean {
  return lazyEvaluateInSource1(chapter) || lazyEvaluateInSource2(chapter)
}

/**
 * Checks whether this chapter number represents
 * Lazy Source chapter 1 (functions, arithmetic,
 * conditionals).
 *
 * @param chapter The chapter number.
 */
export function lazyEvaluateInSource1(chapter: number): boolean {
  return chapter === LAZY_SOURCE_1
}

/**
 * Checks whether this chapter number represents
 * Lazy Source chapter 2 (pairs and lists).
 *
 * @param chapter The chapter number.
 */
export function lazyEvaluateInSource2(chapter: number): boolean {
  return chapter === LAZY_SOURCE_2
}
