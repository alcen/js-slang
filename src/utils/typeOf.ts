import { Value } from "../types"

// We need to define our own typeof in order for null/array to display properly in error messages
export const typeOf = (v: Value) => {
  if (v === null) {
    return 'null'
  } else if (Array.isArray(v)) {
    return 'array'
  } else {
    return typeof v
  }
}
