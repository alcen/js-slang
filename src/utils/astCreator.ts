import * as es from 'estree'
import { astThunkNativeTag } from '../stdlib/lazy'
import { AllowedDeclarations, BlockExpression, FunctionDeclarationExpression } from '../types'
import { typeOf } from './typeOf'

export const locationDummyNode = (line: number, column: number) =>
  literal('Dummy', { start: { line, column }, end: { line, column } })

export const identifier = (name: string): es.Identifier => ({
  type: 'Identifier',
  name
})

/**
 * Constructs a new Thunk value for lazy evaluation. This function
 * does this by taking the original Literal in the Abstract Syntax
 * Tree, and changing it into an ArrowFunctionExpression with the
 * literal in its body
 * @param node The Literal to be transformed.
 */
export const mutateToThunk = (node: es.Literal): es.ObjectExpression => {
  // creates a new copy of node.loc
  function copyLoc() {
    if (!node.loc) {
      return undefined
    } else {
      return {
        start: Object.assign({}, node.loc.start),
        end: Object.assign({}, node.loc.end)
      }
    }
  }
  // make a new object for the new Thunk
  const newNode = node as any
  newNode.type = 'ObjectExpression'
  // get type of literal
  const type = typeOf(node.value)
  // ensure that the raw type becomes a string, not a variable
  const typeRaw = '"' + type + '"'
  // get rid of old value and old raw
  const oldValue = newNode.value
  newNode.value = undefined
  const oldRaw = newNode.raw
  newNode.raw = undefined
  newNode.properties = []
  // set Thunk properties
  // first, create type property
  newNode.properties[0] = {
    type: 'Property',
    start: newNode.start,
    end: newNode.end,
    loc: copyLoc(),
    method: false,
    shorthand: false,
    computed: false,
    key: {
      type: 'Identifier',
      start: newNode.start,
      end: newNode.end,
      loc: copyLoc(),
      name: 'type'
    },
    // value of type is a string literal
    value: {
      type: 'Literal',
      start: newNode.start,
      end: newNode.end,
      loc: copyLoc(),
      value: type,
      raw: typeRaw
    },
    kind: 'init'
  }
  // get arrow function
  const arrowFunction = {
    type: 'ArrowFunctionExpression',
    start: newNode.start,
    end: newNode.end,
    loc: copyLoc(),
    id: null,
    expression: true,
    generator: false,
    params: [],
    body: {
      type: 'Literal',
      start: newNode.start,
      end: newNode.end,
      loc: copyLoc(),
      value: oldValue,
      raw: oldRaw
    }
  }
  // then, create the value property
  // also create lambda function storing original literal
  newNode.properties[1] = {
    type: 'Property',
    start: newNode.start,
    end: newNode.end,
    loc: copyLoc(),
    method: false,
    shorthand: false,
    computed: false,
    key: {
      type: 'Identifier',
      start: newNode.start,
      end: newNode.end,
      loc: copyLoc(),
      name: 'value'
    },
    // value of 'value' is an ArrowFunctionExpression
    value: arrowFunction,
    kind: 'init'
  }
  // get toString to show the normal result
  const toString = {
    type: 'ArrowFunctionExpression',
    start: newNode.start,
    end: newNode.end,
    loc: copyLoc(),
    id: null,
    expression: true,
    generator: false,
    params: [],
    body: {
      type: 'Literal',
      start: newNode.start,
      end: newNode.end,
      loc: copyLoc(),
      value: JSON.stringify(oldValue),
      raw: JSON.stringify(oldRaw)
    },
    // add tag to prevent toString() from getting wrapped in
    // wrapArrowFunctionsToAllowNormalCallsAndNiceToString
    // (toString is already giving it a nice string representation)
    tag: astThunkNativeTag
  }
  // lastly create the toString property
  // so thunks appear as normal values
  newNode.properties[2] = {
    type: 'Property',
    start: newNode.start,
    end: newNode.end,
    loc: copyLoc(),
    method: false,
    shorthand: false,
    computed: false,
    key: {
      type: 'Identifier',
      start: newNode.start,
      end: newNode.end,
      loc: copyLoc(),
      name: 'toString'
    },
    value: toString,
    kind: 'init'
  }
  return newNode
}

export const literal = (value: string | number | boolean, loc?: es.SourceLocation): es.Literal => ({
  type: 'Literal',
  value,
  loc
})

export const memberExpression = (
  object: es.Expression,
  propertyString: string
): es.MemberExpression => ({
  type: 'MemberExpression',
  object,
  computed: false,
  property: identifier(propertyString)
})

export const declaration = (
  name: string,
  kind: AllowedDeclarations,
  init: es.Expression,
  loc?: es.SourceLocation
): es.VariableDeclaration => ({
  type: 'VariableDeclaration',
  declarations: [
    {
      type: 'VariableDeclarator',
      id: identifier(name),
      init
    }
  ],
  kind,
  loc
})

export const constantDeclaration = (name: string, init: es.Expression, loc?: es.SourceLocation) =>
  declaration(name, 'const', init, loc)

export const callExpression = (
  callee: es.Expression,
  args: es.Expression[],
  loc?: es.SourceLocation
): es.CallExpression => ({
  type: 'CallExpression',
  callee,
  arguments: args,
  loc
})

export const expressionStatement = (expression: es.Expression): es.ExpressionStatement => ({
  type: 'ExpressionStatement',
  expression
})

export const blockArrowFunction = (
  params: es.Identifier[],
  body: es.Statement[] | es.BlockStatement,
  loc?: es.SourceLocation
): es.ArrowFunctionExpression => ({
  type: 'ArrowFunctionExpression',
  expression: false,
  generator: false,
  params,
  body: Array.isArray(body) ? blockStatement(body) : body,
  loc
})

export const functionExpression = (
  params: es.Identifier[],
  body: es.Statement[] | es.BlockStatement,
  loc?: es.SourceLocation
): es.FunctionExpression => ({
  type: 'FunctionExpression',
  id: null,
  async: false,
  generator: false,
  params,
  body: Array.isArray(body) ? blockStatement(body) : body,
  loc
})

export const blockStatement = (body: es.Statement[]): es.BlockStatement => ({
  type: 'BlockStatement',
  body
})

export const program = (body: es.Statement[]): es.Program => ({
  type: 'Program',
  sourceType: 'module',
  body
})

export const returnStatement = (
  argument: es.Expression,
  loc?: es.SourceLocation
): es.ReturnStatement => ({
  type: 'ReturnStatement',
  argument,
  loc
})

export const property = (key: string, value: es.Expression): es.Property => ({
  type: 'Property',
  method: false,
  shorthand: false,
  computed: false,
  key: identifier(key),
  value,
  kind: 'init'
})

export const objectExpression = (properties: es.Property[]): es.ObjectExpression => ({
  type: 'ObjectExpression',
  properties
})

export const mutateToCallExpression = (
  node: es.Node,
  callee: es.Expression,
  args: es.Expression[]
) => {
  node.type = 'CallExpression'
  node = node as es.CallExpression
  node.callee = callee
  node.arguments = args
}

export const logicalExpression = (
  operator: es.LogicalOperator,
  left: es.Expression,
  right: es.Expression,
  loc?: es.SourceLocation
): es.LogicalExpression => ({
  type: 'LogicalExpression',
  operator,
  left,
  right,
  loc
})

export const conditionalExpression = (
  test: es.Expression,
  consequent: es.Expression,
  alternate: es.Expression,
  loc?: es.SourceLocation
): es.ConditionalExpression => ({
  type: 'ConditionalExpression',
  test,
  consequent,
  alternate,
  loc
})

export const arrayExpression = (elements: es.Expression[]): es.ArrayExpression => ({
  type: 'ArrayExpression',
  elements
})

export const assignmentExpression = (
  left: es.Identifier | es.MemberExpression,
  right: es.Expression
): es.AssignmentExpression => ({
  type: 'AssignmentExpression',
  operator: '=',
  left,
  right
})

export const binaryExpression = (
  operator: es.BinaryOperator,
  left: es.Expression,
  right: es.Expression,
  loc?: es.SourceLocation
): es.BinaryExpression => ({
  type: 'BinaryExpression',
  operator,
  left,
  right,
  loc
})

export const unaryExpression = (
  operator: es.UnaryOperator,
  argument: es.Expression,
  loc?: es.SourceLocation
): es.UnaryExpression => ({
  type: 'UnaryExpression',
  operator,
  prefix: true,
  argument,
  loc
})

// primitive: undefined is a possible value
export const primitive = (value: any): es.Expression => {
  return value === undefined ? identifier('undefined') : literal(value)
}

export const functionDeclarationExpression = (
  id: es.Identifier,
  params: es.Pattern[],
  body: es.BlockStatement,
  loc?: es.SourceLocation
): FunctionDeclarationExpression => ({
  type: 'FunctionExpression',
  id,
  params,
  body,
  loc
})

export const functionDeclaration = (
  id: es.Identifier | null,
  params: es.Pattern[],
  body: es.BlockStatement,
  loc?: es.SourceLocation
): es.FunctionDeclaration => ({
  type: 'FunctionDeclaration',
  id,
  params,
  body,
  loc
})

export const blockExpression = (
  body: es.Statement[],
  loc?: es.SourceLocation
): BlockExpression => ({
  type: 'BlockExpression',
  body,
  loc
})

export const arrowFunctionExpression = (
  params: es.Pattern[],
  body: es.Expression | es.BlockStatement,
  loc?: es.SourceLocation
): es.ArrowFunctionExpression => ({
  type: 'ArrowFunctionExpression',
  expression: body.type !== 'BlockStatement',
  generator: false,
  params,
  body,
  loc
})

export const variableDeclaration = (
  declarations: es.VariableDeclarator[],
  loc?: es.SourceLocation
): es.VariableDeclaration => ({
  type: 'VariableDeclaration',
  kind: 'const',
  declarations,
  loc
})

export const variableDeclarator = (
  id: es.Pattern,
  init: es.Expression,
  loc?: es.SourceLocation
): es.VariableDeclarator => ({
  type: 'VariableDeclarator',
  id,
  init,
  loc
})

export const ifStatement = (
  test: es.Expression,
  consequent: es.BlockStatement,
  alternate: es.Statement,
  loc?: es.SourceLocation
): es.IfStatement => ({
  type: 'IfStatement',
  test,
  consequent,
  alternate,
  loc
})
