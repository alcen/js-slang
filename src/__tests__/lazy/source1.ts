import { stripIndent } from '../../utils/formatters'
import { expectResult, expectParsedError, expectDisplayResult } from '../../utils/testing'
import { LAZY_SOURCE_1 } from '../../lazyContext'
import { ExecutionMethod } from '../../types'

// ================= UTILITY FUNCTIONS =================
const runTestSuccess = (
  executionMethod: ExecutionMethod,
  program: string,
  result: string,
  testName: string
) =>
  test(testName + ' (' + executionMethod + ')', () => {
    return expectResult(stripIndent(program), {
      chapter: LAZY_SOURCE_1,
      executionMethod
    }).toMatchInlineSnapshot(result)
  })

const runTestError = (
  executionMethod: ExecutionMethod,
  program: string,
  error: string,
  testName: string
) =>
  test(testName + ' (' + executionMethod + ')', () => {
    return expectParsedError(stripIndent(program), {
      chapter: LAZY_SOURCE_1,
      executionMethod
    }).toMatchInlineSnapshot(error)
  })

const runTestForDisplay = (
  executionMethod: ExecutionMethod,
  program: string,
  displayed: string,
  testName: string
) =>
  test(testName + ' (' + executionMethod + ')', () => {
    return expectDisplayResult(stripIndent(program), {
      chapter: LAZY_SOURCE_1,
      executionMethod
    }).toMatchInlineSnapshot(displayed)
  })

// ==================== LAZY TESTS =====================
const isNumberReturnsTrue = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
  is_number(145);
`,
    'true',
    'is number returns true for number'
  )

const isNumberReturnsFalse = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
  is_number(true);
`,
    'false',
    'is number returns false for boolean'
  )

const isBooleanReturnsTrue = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
  is_boolean(true || false);
`,
    'true',
    'is boolean returns true for boolean'
  )

const isBooleanReturnsFalse = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
  is_boolean("hello");
`,
    'false',
    'is boolean returns false for string'
  )

const isStringReturnsTrue = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
  is_string("");
`,
    'true',
    'is string returns true for string'
  )

const isStringReturnsFalse = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
  is_string(() => "world");
`,
    'false',
    'is string returns false for function'
  )

const isFunctionReturnsTrue = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
  is_function((x, y) => x + y);
`,
    'true',
    'is function returns true for function'
  )

const isFunctionReturnsFalse = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
  is_function(670);
`,
    'false',
    'is function returns false for number'
  )

const lazyAccessOfNames = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
  const a = 1 + a;
`,
    'undefined',
    'names are lazily accessed'
  )

const lazyFunctionArguments = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
  function try_me(test, alternative) {
      return test ? 123 : alternative;
  }
  try_me(true, head(null));
`,
    '123',
    'function arguments are lazily evaluated'
  )

const errorEvaluatesToError = (executionMethod: ExecutionMethod) =>
  runTestError(
    executionMethod,
    `
  error("This is the error");
  `,
    '"Line 1: Error: \\"This is the error\\""',
    'error evaluates to error'
  )

const lazyEvaluationOfFunctions = (executionMethod: ExecutionMethod) =>
  runTestForDisplay(
    executionMethod,
    `
  function square(x) {
      display("squared");
      return x * x;
  }
  const sq = square(7);
`,
    'Array []',
    'function does not run if it does not need to'
  )

const lazyMemoisation = (executionMethod: ExecutionMethod) =>
  runTestForDisplay(
    executionMethod,
    `
  function square(x) {
      display("squared");
      return x * x;
  }
  const sq = square(7);
  force(sq);
  force(sq);
  sq;
`,
    `
Array [
  "\\"squared\\"",
]
`,
    'function only runs once and value is memoised'
  )

// ================== RUN THE TESTS ====================
const testArray = [
  isNumberReturnsTrue,
  isNumberReturnsFalse,
  isBooleanReturnsTrue,
  isBooleanReturnsFalse,
  isStringReturnsTrue,
  isStringReturnsFalse,
  isFunctionReturnsTrue,
  isFunctionReturnsFalse,
  lazyAccessOfNames,
  lazyFunctionArguments,
  lazyEvaluationOfFunctions,
  lazyMemoisation,
  errorEvaluatesToError
]
testArray.forEach(tst => tst('interpreter'))
testArray.forEach(tst => tst('native'))
testArray.forEach(tst => tst('auto'))
