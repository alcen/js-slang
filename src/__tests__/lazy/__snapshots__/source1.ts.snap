// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error evaluates to error (auto): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "error(\\"This is the error\\");",
  "displayResult": Array [],
  "errors": Array [
    ExceptionError {
      "error": [Error: "This is the error"],
      "location": SourceLocation {
        "end": Position {
          "column": 26,
          "line": 1,
        },
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "severity": "Error",
      "type": "Runtime",
    },
  ],
  "parsedErrors": "Line 1: Error: \\"This is the error\\"",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`error evaluates to error (interpreter): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "error(\\"This is the error\\");",
  "displayResult": Array [],
  "errors": Array [
    ExceptionError {
      "error": [Error: "This is the error"],
      "location": SourceLocation {
        "end": Position {
          "column": 26,
          "line": 1,
        },
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "severity": "Error",
      "type": "Runtime",
    },
  ],
  "parsedErrors": "Line 1: Error: \\"This is the error\\"",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`error evaluates to error (native): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "error(\\"This is the error\\");",
  "displayResult": Array [],
  "errors": Array [
    ExceptionError {
      "error": [Error: "This is the error"],
      "location": SourceLocation {
        "end": Position {
          "column": 26,
          "line": 1,
        },
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "severity": "Error",
      "type": "Runtime",
    },
  ],
  "parsedErrors": "Line 1: Error: \\"This is the error\\"",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`function arguments are lazily evaluated (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function try_me(test, alternative) {
    return test ? 123 : alternative;
}
try_me(true, head(null));",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 123,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`function arguments are lazily evaluated (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function try_me(test, alternative) {
    return test ? 123 : alternative;
}
try_me(true, head(null));",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 123,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`function arguments are lazily evaluated (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function try_me(test, alternative) {
    return test ? 123 : alternative;
}
try_me(true, head(null));",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 123,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`function does not run if it does not need to (auto): expectDisplayResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function square(x) {
    display(\\"squared\\");
    return x * x;
}
const sq = square(7);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": undefined,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`function does not run if it does not need to (interpreter): expectDisplayResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function square(x) {
    display(\\"squared\\");
    return x * x;
}
const sq = square(7);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": undefined,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`function does not run if it does not need to (native): expectDisplayResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function square(x) {
    display(\\"squared\\");
    return x * x;
}
const sq = square(7);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": undefined,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`function only runs once and value is memoised (auto): expectDisplayResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function square(x) {
    display(\\"squared\\");
    return x * x;
}
const sq = square(7);
force(sq);
force(sq);
sq;",
  "displayResult": Array [
    "\\"squared\\"",
  ],
  "errors": Array [],
  "parsedErrors": "",
  "result": 49,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`function only runs once and value is memoised (interpreter): expectDisplayResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function square(x) {
    display(\\"squared\\");
    return x * x;
}
const sq = square(7);
force(sq);
force(sq);
sq;",
  "displayResult": Array [
    "\\"squared\\"",
  ],
  "errors": Array [],
  "parsedErrors": "",
  "result": 49,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`function only runs once and value is memoised (native): expectDisplayResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function square(x) {
    display(\\"squared\\");
    return x * x;
}
const sq = square(7);
force(sq);
force(sq);
sq;",
  "displayResult": Array [
    "\\"squared\\"",
  ],
  "errors": Array [],
  "parsedErrors": "",
  "result": 49,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is boolean returns false for string (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_boolean(\\"hello\\");",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is boolean returns false for string (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_boolean(\\"hello\\");",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is boolean returns false for string (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_boolean(\\"hello\\");",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is boolean returns true for boolean (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_boolean(true || false);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is boolean returns true for boolean (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_boolean(true || false);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is boolean returns true for boolean (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_boolean(true || false);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is function returns false for number (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_function(670);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is function returns false for number (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_function(670);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is function returns false for number (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_function(670);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is function returns true for function (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_function((x, y) => x + y);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is function returns true for function (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_function((x, y) => x + y);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is function returns true for function (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_function((x, y) => x + y);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is number returns false for boolean (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_number(true);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is number returns false for boolean (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_number(true);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is number returns false for boolean (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_number(true);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is number returns true for number (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_number(145);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is number returns true for number (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_number(145);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is number returns true for number (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_number(145);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is string returns false for function (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_string(() => \\"world\\");",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is string returns false for function (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_string(() => \\"world\\");",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is string returns false for function (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_string(() => \\"world\\");",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is string returns true for string (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_string(\\"\\");",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is string returns true for string (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_string(\\"\\");",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is string returns true for string (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_string(\\"\\");",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`names are lazily accessed (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const a = 1 + a;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": undefined,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`names are lazily accessed (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const a = 1 + a;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": undefined,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`names are lazily accessed (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const a = 1 + a;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": undefined,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;
